### PR TITLE
Categorical_crossentropy() performance gain

### PR DESCRIPTION
**Pull request by PierrickPochelu**
This rescale is generally useless because reduce_sum is generally equal to 1. Disabling the rescale is significantly faster (E.g, from 11 seconds to 7 seconds).

Why reduced_sum(output) is always equal to 1 ? Because the rescaling is performed on values WHICH ARE NOT A LOGIT. In general, the ouput has already been already rescaled with a softmax.
 
 I choose to enable / disable with the default behaviour is forced_rescale=True which is the previous behaviour. The default behaviour is preserved for safety. I recommend to put forced_rescale=False later.

Example of code to compare:
```
import numpy as np
from numpy.ma.core import _frommethod
from tensorflow.keras.activations import softmax
from tensorflow import constant
from keras.backend import categorical_crossentropy
NB_SAMPLES=1000
NB_CLASSES=1000
y=np.zeros((NB_SAMPLES,NB_CLASSES))
y[:,0]=1
y_pred=np.random.uniform(0,1,(NB_SAMPLES,NB_CLASSES))
y_pred=softmax(constant(y_pred)).numpy()
tf_y=constant(y)
tf_y_pred=constant(y_pred)


import time
st_time=time.time()
for i in range(1000):
  res=categorical_crossentropy(tf_y, tf_y_pred ) # old behaviour
print("Enlapsed time:", time.time()-st_time)
print("Loss:", np.mean(res))

st_time=time.time()
for i in range(1000):
  res=categorical_crossentropy(tf_y, tf_y_pred,forced_rescale=False) # new behaviour
print("Enlapsed time:", time.time()-st_time)
print("Loss:", np.mean(res))
```

returns:

```
Enlapsed time: 11.447702169418335
Loss: 6.944476480206596
Enlapsed time: 7.416125059127808
Loss: 6.944476480206596
```
You observe the computed loss is strictly the same.

### CODE REVIEW

**Comment by divyashreepathihalli** [comment]
@PierrickPochelu can you please add the unit tests to cover the edge cases discussed here - https://github.com/keras-team/keras/issues/17102

File: keras/backend.py [hunk 1]
```diff
@@ -5484,7 +5484,7 @@ def _get_logits(output, from_logits, op_type, fn_name):
 @keras_export("keras.backend.categorical_crossentropy")
 @tf.__internal__.dispatch.add_dispatch_support
 @doc_controls.do_not_generate_docs
-def categorical_crossentropy(target, output, from_logits=False, axis=-1):
+def categorical_crossentropy(target, output, from_logits=False, forced_rescale=True, axis=-1):
```
Comment (divyashreepathihalli) [inline keras/backend.py:5487]: with `from_logits` set to `False`, the output is already normalized. So this additional parameter is unnecessary

File: keras/backend.py [hunk 2]
```diff
@@ -5494,6 +5494,8 @@ def categorical_crossentropy(target, output, from_logits=False, axis=-1):
             case `output` is expected to be the logits).
         from_logits: Boolean, whether `output` is the
             result of a softmax, or is a tensor of logits.
+        forced_rescale: Boolean, whether the `output` is rescaled
```
Comment (divyashreepathihalli) [inline keras/backend.py:5497]: you can remove this

File: keras/backend.py [hunk 3]
```diff
@@ -5540,7 +5542,9 @@ def categorical_crossentropy(target, output, from_logits=False, axis=-1):
         )
 
     # scale preds so that the class probas of each sample sum to 1
-    output = output / tf.reduce_sum(output, axis, True)
+    if forced_rescale:
```
Comment (divyashreepathihalli) [inline keras/backend.py:5545]: this conditional check can be removed with the added arg

**Review by divyashreepathihalli — CHANGES_REQUESTED** [review]
Thank you for the PR! 
Few minor comments
- So the first edge case where all outputs are zero is not a feasible edge case in reality, it can be ignored
- please add the unit tests for single class edge case
- Upon removing this line, the unit tests need to be updated

File: keras/backend.py (hunk 3 above)
Comment (fchollet) [reply keras/backend.py:5545]: Agree, we can remove the arg and the conditional. One thing to note is that the change might be backwards incompatible for folks who mistakenly had non-softmax inputs to the function without passing `from_logits=True`. We need to be careful with the rollout.

**Comment by PierrickPochelu** [comment]
Agree, I removed the arg and the conditional in my last commit.

**Review by divyashreepathihalli — CHANGES_REQUESTED** [review]
@PierrickPochelu, you would still need to update the test file. There are some failing tests - more information here - https://source.cloud.google.com/results/invocations/0c68e47a-876c-4d01-9689-340466310029/targets/keras%2Fpip%2Fpresubmit/log

**Comment by PierrickPochelu** [comment]
I have seen some logical mistakes in the unit tests.

This same wrong input have been used in multiple unit tests : [0.5, 0.89, 0.6] in combination of from_logits=False .

Assuming a correct usage of the API, we should have:
- either categorical_crossentropy(Y, [0.**0**5, 0.89, 0.**0**6], from_logits=False)
- or categorical_crossentropy(Y, [0.5, 0.89, 0.6], from_logits=**True** )

I send a PR of the unit tests ASAP

**Comment by divyashreepathihalli** [comment]
> The 8 wrong test units have been corrected.

@PierrickPochelu  can you also add the unit tests for single class edge case?

File: keras/backend.py (hunk 3 above)
Comment (divyashreepathihalli) [reply keras/backend.py]: @fchollet - we can add the check and raise a warning here - https://github.com/keras-team/keras/blob/master/keras/backend.py#:~:text=def%20_get_logits(output%2C%20from_logits%2C%20op_type%2C%20fn_name)%3A 
wdyt?

**Comment by PierrickPochelu** [comment]
Any other edge cases to handle ?

**Comment by fchollet** [comment]
One drive-by comment. If you care about speed, you should use a tf.function. Something like

```
import tensorflow as tf
import numpy as np

from tensorflow.keras.activations import softmax
from tensorflow import constant
from keras.backend import categorical_crossentropy

NB_SAMPLES=1000
NB_CLASSES=1000
y = np.zeros((NB_SAMPLES,NB_CLASSES))
y[:,0] = 1
y_pred = np.random.uniform(0,1,(NB_SAMPLES,NB_CLASSES))

y = tf.constant(y)
y_pred = tf.constant(y_pred)

@tf.function
def fn(y, y_pred):
  y_pred = softmax(y_pred)
  return categorical_crossentropy(y, tf_y_pred)

fn(y, y_pred) # warmup

import time
st_time = time.time()
for i in range(1000):
  fn(y, y_pred)
print("Enlapsed time:", time.time() - st_time)
```

It should be about 3x faster.

Also note that it will use `from_logits` automatically.

**Comment by divyashreepathihalli** [comment]
> 

I have added another comment here.- https://github.com/keras-team/keras/pull/17140#discussion_r1002211748

**Comment by divyashreepathihalli** [comment]
@PierrickPochelu can you please fix the lint errors - https://github.com/keras-team/keras/actions/runs/3329674806/jobs/5513349190

**Comment by PierrickPochelu** [comment]
Done

**Comment by divyashreepathihalli** [comment]
Hey @PierrickPochelu sorry about the late response. 
Can you please resolve these errors 
- https://source.cloud.google.com/results/invocations/e76b488f-8bef-4700-b5c5-9b3d8e164502
- https://source.cloud.google.com/results/invocations/0d3b0b95-d078-42b9-bed5-7d136f630013

**Comment by PierrickPochelu** [comment]
All CI tests are OK

**Review by fchollet — APPROVED** [review]
LGTM, thanks! Let's try to merge. Please be mindful of the fact that we might end up being unable to merge due to backwards compatibility breakage. This is an attempt.

**Comment by divyashreepathihalli** [comment]
@PierrickPochelu, this change breaks several tests internally and there are no straight forward fixes. Since the  performance improvement is not very significant to break backwards compatibility, we will not be merging this changes.
